File: service/adapter/AdapterFactory.js
const adapters = {};

function registerAdapter(name, adapterClass) {
  adapters[name] = adapterClass;
}

function createAdapter(name, options) {
  const Adapter = adapters[name];
  if (!Adapter) throw new Error(`Adapter ${name} not registered`);
  return new Adapter(options);
}

module.exports = { registerAdapter, createAdapter };

File: dependency-mapper/dependency-mapper/src/main/java/com/example/mapper/adapter/AdapterFactory.java
package com.example.mapper.adapter;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public class AdapterFactory {
    private final Map<String, Supplier<DataSourceAdapter<?>>> registry = new HashMap<>();

    public void registerAdapter(String name, Supplier<DataSourceAdapter<?>> supplier) {
        registry.put(name, supplier);
    }

    public DataSourceAdapter<?> createAdapter(String name) {
        Supplier<DataSourceAdapter<?>> supplier = registry.get(name);
        if (supplier == null) {
            throw new IllegalArgumentException("Adapter " + name + " not registered");
        }
        return supplier.get();
    }
}
